### PR TITLE
fix handling of failed watches that result in snapshot

### PR DESCRIPTION
**Pull request by eyakubovich**
This also fixes the cursor not being used in some cases.

### CODE REVIEW

File: subnet/watch.go [hunk 1]
```diff
@@ -23,9 +23,10 @@ import (
 
 func WatchLeases(ctx context.Context, sm Manager, network string, receiver chan []Event) {
 	lw := &leaseWatcher{}
+	var cursor interface{}
```
Comment (jonboulle) [inline subnet/watch.go]: Gross, can't you use -1 instead of an empty interface?

Comment (eyakubovich) [inline subnet/watch.go]: The idea is that the cursor is opaque to these "helper methods" and at the interface level. Each subnet.Manager impl is free to implement the cursor as it sees fit.

File: subnet/watch.go [hunk 2]
```diff
@@ -21,11 +21,16 @@ import (
 	"github.com/coreos/flannel/Godeps/_workspace/src/golang.org/x/net/context"
 )
 
+// WatchLeases performs a long term watch of the given network's subnet leases
+// and communicates addition/deletion events on receiver channel. It takes care
+// of handling "fall-behind" logic where the history window has advanced to far
```
Comment (jonboulle) [inline subnet/watch.go]: too far

File: subnet/watch.go [hunk 3]
```diff
@@ -21,11 +21,16 @@ import (
...
+// and it needs diff the latest snapshot with its saved state and generate events
```
Comment (jonboulle) [inline subnet/watch.go]: to diff

**Comment by jonboulle** [comment]
lgtm

**Pull request merged**